### PR TITLE
CORE-15694: Update to Java 17

### PR DESCRIPTION
**Pull request by tlawson3**
Tested via https://ci02.dev.r3.com/job/QA/job/Craft_Tests/job/qa-regpack-csde/job/main/344/

### CODE REVIEW

File: build.gradle
Comment (marvincudjoe) [inline build.gradle:1]: `javaVersion` doesn't look to be in use. Could we add the[ java tool chain](https://kotlinlang.org/docs/gradle-configure-project.html#gradle-java-toolchains-support).
IntelliJ should then pick up the "language support" as 17. Gradle tasks should also pick up 17 as the default

Comment (tlawson3) [reply build.gradle:1]: So within the allprojects block have this?
```
java {
        toolchain {
            languageVersion = JavaLanguageVersion.of(17)
        }
    }
```

@corda/blt do you have any advice?

Comment (marvincudjoe) [reply build.gradle:1]: I've generally seen it outside a specific block, just on its own. Some build scripts might have `sourceCompatibility` but the [recommendation](https://docs.gradle.org/current/userguide/java_plugin.html#toolchain_and_compatibility) is to use the toolchain

**Review by yarivus — APPROVED** [review]
LGTM

File: build.gradle
Comment (driessamyn) [reply build.gradle:1]: > `javaVersion` doesn't look to be in use. Could we add the[ java tool chain](https://kotlinlang.org/docs/gradle-configure-project.html#gradle-java-toolchains-support). IntelliJ should then pick up the "language support" as 17. Gradle tasks should also pick up 17 as the default

AFAIK, this is used by the Jenkins pipeline to make sure a JAVA 17 JDK is available in the Jenkins docker image, for compilation.
This is separate from the Gradle JAVA 17 toolchain feature. Off course we wouldn't want to download a fresh SDK on each build.

Comment (wzur-r3) [reply build.gradle:1]: `javaVersion` is used by Jenkins to provide relevant build environment, in this case with JDK17 as the default JDK.

It has nothing to do with Gradle.

Comment (tlawson3) [reply build.gradle:1]: The discussion is about line https://github.com/corda/CSDE-cordapp-template-java/pull/67/files#diff-49a96e7eea8a94af862798a45174e6ac43eb4f8b4bd40759b5da63ba31ec3ef7L16 in the gradle.build file, to be clear.

Comment (tlawson3) [reply build.gradle:1]: Found an example of another project using toolchain to determine the Java version: https://github.com/corda/corda-api/blob/bba7588fbb01eea0bf452cbcca1ebd5cb13a623e/build.gradle#L124-L126

Found an example of not using toolchain, but using source/target compatibility: https://github.com/corda/corda/blob/eccb9b4af63fe74e596428f1594eefb2e2fd2505/build.gradle#L290-L291

I don't know which is better, and would either of these require a change to the kotlin project to keep things equivalent?

**Review by marvincudjoe — APPROVED** [review]

**Review by yarivus — APPROVED** [review]
LGTM

**Comment by tlawson3** [comment]
Tested again to be sure: https://ci02.dev.r3.com/job/QA/job/Craft_Tests/job/qa-regpack-csde/job/main/349/

**Pull request merged**